apps/server/lib/lexical/server/configuration/support.ex

```diff
@@ -1,65 +1,65 @@
 defmodule Lexical.Server.Configuration.Support do
-  alias Lexical.Protocol.Types.ClientCapabilities
-
-  defstruct code_action_dynamic_registration?: false,
-            hierarchical_document_symbols?: false,
-            snippet?: false,
-            deprecated?: false,
-            tags?: false,
-            signature_help?: false,
-            work_done_progress?: false
-
-  def new(%ClientCapabilities{} = client_capabilities) do
-    dynamic_registration? =
-      client_capabilities
-      |> get_in([:text_document, :code_action, :dynamic_registration])
-      |> bool()
+  @moduledoc false
 
-    hierarchical_symbols? =
-      client_capabilities
-      |> get_in([:text_document, :document_symbol, :hierarchical_document_symbol_support])
-      |> bool()
-
-    snippet? =
-      client_capabilities
-      |> get_in([:text_document, :completion, :completion_item, :snippet_support])
-      |> bool()
+  alias Lexical.Protocol.Types.ClientCapabilities
 
-    deprecated? =
-      client_capabilities
-      |> get_in([:text_document, :completion, :completion_item, :deprecated_support])
-      |> bool()
+  # To track a new client capability, add a new field and the path to the
+  # capability in the `Lexical.Protocol.Types.ClientCapabilities` struct
+  # to this mapping:
+  @client_capability_mapping [
+    code_action_dynamic_registration: [
+      :text_document,
+      :code_action,
+      :dynamic_registration
+    ],
+    hierarchical_symbols: [
+      :text_document,
+      :document_symbol,
+      :hierarchical_document_symbol_support
+    ],
+    snippet: [
+      :text_document,
+      :completion,
+      :completion_item,
+      :snippet_support
+    ],
+    deprecated: [
+      :text_document,
+      :completion,
+      :completion_item,
+      :deprecated_support
+    ],
+    tags: [
+      :text_document,
+      :completion,
+      :completion_item,
+      :tag_support
+    ],
+    signature_help: [
+      :text_document,
+      :signature_help
+    ],
+    work_done_progress: [
+      :window,
+      :work_done_progress
+    ]
+  ]
 
-    tags? =
-      client_capabilities
-      |> get_in([:text_document, :completion, :completion_item, :tag_support])
-      |> bool()
+  defstruct Enum.map(@client_capability_mapping, fn {field, _} -> {field, false} end)
```

Comment:
i usually like to see the defstruct more spelled out than built.

Reply:
I agree in the usual case. In this case, it seemed like a lot of duplication.

The alternative might be to write things out here explicitly, and then make it so that it errors in calls to `new` if there's any mismatch between the `@client_capability_mapping` and the fields in the struct.

Reply:
I decided to just write it out. It's not that much, and any mismatch is guarded by the `struct(__MODULE__, defaults)` call.